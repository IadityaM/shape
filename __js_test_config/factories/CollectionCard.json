{"data"=>{"id"=>"2877250", "type"=>"collection_cards", "attributes"=>{"class_type"=>"CollectionCard::Primary", "order"=>0, "width"=>1, "height"=>1, "parent_id"=>762833, "pinned"=>false, "image_contain"=>false, "card_question_type"=>nil, "is_cover"=>false, "filter"=>"transparent_gray", "hidden"=>false, "show_replace"=>true, "updated_at"=>"2019-12-16T20:17:09.339Z", "col"=>1, "row"=>1, "section_type"=>nil, "idea_id"=>nil, "link"=>false, "pinned_and_locked"=>false, "can_edit_parent"=>false, "is_master_template_card"=>false}, "relationships"=>{"item"=>{"meta"=>{"included"=>false}}, "collection"=>{"meta"=>{"included"=>false}}, "parent"=>{"meta"=>{"included"=>false}}, "record"=>{"meta"=>{"included"=>false}}, "templated_from"=>{"meta"=>{"included"=>false}}}}}