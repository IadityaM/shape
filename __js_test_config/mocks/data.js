export const fakeTextItemAttrs = {
  id: 1,
  type: 'Item::TextItem',
  name: 'My Cool Item',
  text_data: 'This is the content for the item and it contains multiple sentences. Like this one.',
  breadcrumb: [['collections', 1, 'Some collection'], ['items', 1, 'my item']],
  parentPath: '/',
}
export const fakeTextItem = {
  ...fakeTextItemAttrs,
  rawAttributes: jest.fn().mockReturnValue(fakeTextItemAttrs),
  getRecordType: jest.fn().mockReturnValue('items'),
  toJS: jest.fn().mockReturnValue(fakeTextItemAttrs),
}
export const fakeVideoItemAttrs = {
  id: 2,
  type: 'Item::VideoItem',
  name: 'Crazy Honey Badger',
  url: 'https://www.youtube.com/watch?v=4r7wHMg5Yjg'
}
export const fakeVideoItem = {
  ...fakeVideoItemAttrs,
  rawAttributes: jest.fn().mockReturnValue(fakeTextItemAttrs),
  getRecordType: jest.fn().mockReturnValue('items'),
}
export const fakeImageItemAttrs = {
  id: 3,
  type: 'Item::ImageItem',
  name: 'Earth from Space',
  filestack_file: {
    id: 1,
    url: 'https://www.nasa.gov/sites/default/files/styles/full_width_feature/public/thumbnails/image/iss052e023801_0.jpg',
  }
}
export const fakeImageItem = {
  ...fakeImageItemAttrs,
  rawAttributes: jest.fn().mockReturnValue(fakeTextItemAttrs),
  getRecordType: jest.fn().mockReturnValue('items'),
}
export const fakeItemCard = {
  id: 10,
  order: 0,
  height: 1,
  width: 1,
  record: fakeTextItem,
  item: fakeTextItem,
}
export const fakeCollectionCard = {
  id: 11,
  order: 1,
  height: 1,
  width: 1,
  record: fakeCollection,
  item: fakeCollection,
  reference: false,
}
export const fakeCollection = {
  id: 1,
  name: 'My Workspace X',
  type: 'Collection',
  breadcrumb: [],
  collection_cards: [
    fakeItemCard, fakeItemCard, fakeItemCard
  ]
}
export const fakePosition = {
  xPos: 0,
  yPos: 0,
  height: 1,
  width: 1,
}
