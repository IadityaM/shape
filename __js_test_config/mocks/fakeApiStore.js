const fakeApiStore = ({ findResult = '', findAllResult = [], requestResult = '' } = {}) => {
  return {
    add: jest.fn().mockName('add'),
    removeAll: jest.fn().mockName('removeAll'),
    find: jest.fn()
      .mockReturnValue(findResult),
    findAll: jest.fn()
      .mockReturnValue(findAllResult),
    request: jest.fn()
      .mockReturnValue(Promise.resolve(requestResult))
      .mockName('request'),
    fetch: jest.fn()
      .mockReturnValue(Promise.resolve())
      .mockName('fetch'),
    sync: jest.fn().mockName('sync'),
    setCurrentUserId: jest.fn(),
    currentUserId: 1,
    currentUser: {
      id: 1,
      current_user_collection_id: 99,
      current_organization: {
        name: 'test org'
      },
      name: 'Johnny Appleseed',
      pic_url_square: 'https://s3.amazonaws.com/pic.png',
      groups: [],
      terms_accepted: true,
    },
    collections: [],
    items: [],
  }
}

export default fakeApiStore
