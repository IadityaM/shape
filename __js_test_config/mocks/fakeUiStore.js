import { observable } from "mobx"

const fakeUiStore = {
  gridSettings: {
    cols: 4,
    gutter: 20,
    gridW: 312,
    gridH: 250,
  },
  defaultGridSettings: {
    cols: 4,
    gutter: 20,
    gridW: 312,
    gridH: 250,
  },
  blankContentToolState: {
    order: null,
    width: null,
    height: null,
    replacingId: null,
    placeholderCard: null,
  },
  dialogConfig: {
    open: null,
    prompt: null,
    onConfirm: null,
    onCancel: null,
    iconName: null,
    confirmText: 'OK',
    cancelText: 'Cancel',
    onClose: jest.fn(),
  },
  activityLogOpen: false,
  activityLogPosition: {
    x: 0,
    y: 0,
    h: 1,
    w: 1,
  },
  activityLogMoving: false,
  cardMenuOpen: {
    id: false,
    x: 0,
    y: 0,
    offsetX: 0,
    offsetY: 0,
  },
  selectedArea: { minX: null, maxX: null, minY: null, maxY: null },
  openBlankContentTool: jest.fn(),
  closeBlankContentTool: jest.fn(),
  setBctPlaceholderCard: jest.fn(),
  closeCardMenu: jest.fn(),
  openCardMenu: jest.fn(),
  closeMoveMenu: jest.fn(),
  resetSelectionAndBCT: jest.fn(),
  resetCardPositions: jest.fn(),
  rolesMenuOpen: false,
  isLoading: false,
  dismissedMoveHelper: false,
  selectedAreaEnabled: false,
  setSelectedArea: jest.fn(),
  selectedCardIds: [],
  selectCardId: jest.fn(),
  deselectCards: jest.fn(),
  setViewingRecord: jest.fn(),
  viewingCollection: null,
  isViewingHomepage: false,
  viewingItem: null,
  movingFromCollectionId: null,
  movingCardIds: [],
  openMoveMenu: jest.fn(),
  update: jest.fn(),
  alert: jest.fn(),
  alertOk: jest.fn(),
  defaultAlertError: jest.fn(),
  confirm: jest.fn(),
  closeDialog: jest.fn(),
  cardAction: 'move',
  blurContent: false,
  organizationMenuPage: 'organizationMenuPage',
  organizationMenuGroupId: null,
  organizationMenuOpen: false,
  expandedThreadKey: null,
  zoomLevel: 1,
  expandThread: jest.fn(),
  openGroup: jest.fn(),
  openOptionalMenus: jest.fn(),
  trackEvent: jest.fn(),
  trackedRecords: {},
  editingName: observable([]),
  activityLogPage: 'comments',
  pageMenuOpen: false,
  searchText: '',
  collectionCardSortOrder: '',
  addNewCard: jest.fn(),
  removeNewCard: jest.fn(),
  isNewCard: jest.fn(),
  autocompleteMenuClosed: jest.fn(),
  captureKeyboardGridClick: jest.fn(),
  popupAlert: jest.fn(),
  stopDragging: jest.fn(),
  multiMoveCardIds: [],
  setSnoozeChecked: jest.fn(),
  scrollToTop: jest.fn(),
  scrollToBottom: jest.fn(),
  scrollToBottomOfComments: jest.fn(),
  scrollToBottomOfModal: jest.fn(),
  popupSnackbar: jest.fn(),
  showPermissionsAlert: jest.fn(),
  gridHeightFor: jest.fn().mockReturnValue(250),
  performActionAfterRoute: jest.fn(),
  linkedBreadcrumbTrailForRecord: jest
    .fn()
    .mockImplementation(x => x.breadcrumb),
  addEmptySpaceClickHandler: jest.fn(),
  removeEmptySpaceClickHandler: jest.fn(),
  setEditingCardCover: jest.fn(),
  adminAudienceMenuOpen: false,
  actionMenuOpenForCard: jest.fn(true).mockImplementation(x => false),
  textMenuOpenForCard: jest.fn(true).mockImplementation(x => false),
  replyingToCommentId: false,
  setReplyingToComment: jest.fn(),
  setCommentingOnRecord: jest.fn(),
  setActivityLogPage: jest.fn(),
  setBodyBackgroundImage: jest.fn(),
  setBodyFontColor: jest.fn(),
  selectedTextRangeForCard: jest.fn(),
  reselectCardIds: jest.fn(),
  reselectOnlyEditableRecords: jest.fn(),
  reselectOnlyMovableCards: jest.fn(),
  setMovingCards: jest.fn(),
  setVisibleCols: jest.fn(),
  setVisibleRows: jest.fn(),
  setPlaceholderSpot: jest.fn(),
  setDroppingFilesCount: jest.fn(),
  drag: jest.fn(),
  startDragging: jest.fn(),
  adjustZoomLevel: jest.fn(),
  zoomIn: jest.fn(),
  zoomOut: jest.fn(),
  determineZoomLevels: jest.fn(),
  maxCols: jest.fn(),
  maxGridWidth: jest.fn().mockReturnValue(1384),
  pageMargins: jest.fn().mockReturnValue({ left: 100, top: 50 }),
  relativeZoomLevel: 2,
  zoomLevels: [],
  isSelected: jest.fn(),
  toggleSelectedCardId: jest.fn(),
  openContextMenu: jest.fn(),
  clearTextEditingItem: jest.fn(),
  positionForCoordinates: jest.fn().mockReturnValue({ x:0, y:0, xPos: 0, yPos: 0, width: 1, height: 1 }),
  dragGridSpot: new Map(),
  droppingFilesCount: 0,
}

export default fakeUiStore
