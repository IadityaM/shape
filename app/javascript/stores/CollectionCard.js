import BaseRecord from './BaseRecord'

class CollectionCard extends BaseRecord {
  record() {
    if (this.item) {
      return this.item
    } else if (this.collection) {
      return this.collection
    }
    return null
  }
}
CollectionCard.type = 'collection_cards'

export default CollectionCard
