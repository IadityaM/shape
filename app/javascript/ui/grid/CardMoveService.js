import { apiStore, uiStore } from '~/stores'

class CardMoveService {
  constructor({ apiStore, uiStore }) {
    this.apiStore = apiStore
    this.uiStore = uiStore
  }

  static async moveCards(placement) {
    return new this({ apiStore, uiStore }).moveCards(placement)
  }

  static moveErrors() {
    return new this({ apiStore, uiStore }).moveErrors()
  }

  async moveCards(placement) {
    const { apiStore, uiStore } = this
    const { viewingCollection, cardAction } = uiStore
    // Viewing collection might not be set, such as on the search page
    if (!viewingCollection) {
      uiStore.alert("You can't move an item here")
      return
    }

    const collectionId = viewingCollection.id
    const movingFromCollection = apiStore.find(
      'collections',
      uiStore.movingFromCollectionId
    )
    const error = this.moveErrors({
      viewingCollection,
      movingFromCollection,
      cardAction,
    })

    if (error) {
      if (!viewingCollection.can_edit_content) {
        uiStore.confirm({
          prompt: error,
          confirmText: 'Continue',
          iconName: 'Alert',
          onConfirm: () => {},
          onCancel: () => {
            uiStore.setMovingCards([])
            uiStore.update('selectedCardIds', [])
          },
        })
      } else {
        uiStore.alert(error)
      }
      return
    }

    const cardIds = [...uiStore.movingCardIds]
    let data = {
      to_id: collectionId,
      from_id: uiStore.movingFromCollectionId,
      collection_card_ids: cardIds,
      placement,
    }

    try {
      uiStore.update('isLoadingMoveAction', true)
      let successMessage
      switch (cardAction) {
        case 'move':
          await apiStore.moveCards(data)
          successMessage = 'Items successfully moved !'
          break
        case 'link':
          await apiStore.linkCards(data)
          successMessage = 'Items successfully linked!'
          break
        case 'duplicate':
          await apiStore.duplicateCards(data)
          successMessage = 'Items successfully duplicated!'
          break
        case 'useTemplate': {
          data = {
            parent_id: data.to_id,
            template_id: data.from_id,
            placement,
          }
          await apiStore.createTemplateInstance(data)
          successMessage = 'Your template instance has been created!'
          break
        }
        default:
          return
      }
      // always refresh the current collection
      // TODO: what if you "moved/duplicated to BOTTOM"?
      await viewingCollection.API_fetchCards()

      uiStore.update('isLoadingMoveAction', false)
      uiStore.popupSnackbar({ message: successMessage })
      uiStore.resetSelectionAndBCT()
      uiStore.closeMoveMenu()
      if (placement === 'beginning') {
        uiStore.scrollToTop()
      } else {
        uiStore.scrollToBottom()
      }
      if (cardAction === 'move') {
        // we actually want to reselect the cards at this point
        uiStore.reselectCardIds(cardIds)
      }
    } catch (e) {
      uiStore.update('isLoadingMoveAction', false)
      let message = 'You cannot move a collection within itself.'
      if (e && e.error && e.error[0]) {
        message = e.error[0]
      }
      uiStore.alert(message)
    }
  }

  moveErrors({ viewingCollection, movingFromCollection, cardAction }) {
    if (!viewingCollection.can_edit_content) {
      return 'You only have view access to this collection. Would you like to keep moving the cards?'
    } else if (
      // don't allow moving cards from templates to non-templates
      cardAction === 'move' &&
      movingFromCollection &&
      movingFromCollection.isMasterTemplate &&
      !viewingCollection.isMasterTemplate
    ) {
      return "You can't move pinned template items out of a template"
    } else if (
      cardAction === 'useTemplate' &&
      viewingCollection.isMasterTemplate
    ) {
      return "You can't create a template instance inside another template. You may be intending to create or duplicate a master template into here instead."
    } else if (
      viewingCollection.isTestCollection ||
      viewingCollection.isTestDesign
    ) {
      return "You can't move cards into a test collection"
    }
    return ''
  }
}

export default CardMoveService
