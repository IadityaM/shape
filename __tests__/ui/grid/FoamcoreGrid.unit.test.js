import CollectionCard from '~/stores/jsonApi/CollectionCard'
import FoamcoreGrid from '~/ui/grid/FoamcoreGrid'
import CardMoveService from '~/utils/CardMoveService'
import fakeApiStore from '#/mocks/fakeApiStore'
import fakeUiStore from '#/mocks/fakeUiStore'
import { fakeCollectionCard, fakeCollection } from '#/mocks/data'
import v from '~/utils/variables'

// because of mdlPlaceholder... without this mock it blows up
jest.mock('../../../app/javascript/stores/jsonApi/CollectionCard')
// also allows us to mock + test that the model constructor was called
CollectionCard.mockImplementation((data, apiStore) => {
  return data
})

let props, wrapper, component, rerender, cards, cardA, cardB, cardC
let idCounter = 0

// NOTE: jest window.innerWidth is 1024px
const jestInnerWidth = 1024

function createCard(data) {
  idCounter += 1
  const id = idCounter.toString()
  return {
    ...fakeCollectionCard,
    ...data,
    id,
    record: { internalType: 'items' },
  }
}

describe('FoamcoreGrid', () => {
  beforeEach(() => {
    cardA = createCard({ row: 1, col: 5 })
    cardB = createCard({ row: 0, col: 1, width: 2, height: 2 })
    cardC = createCard({ row: 2, col: 0, width: 2 })
    const collection = fakeCollection

    collection.num_columns = 16
    collection.cardMatrix = [[], [], []]
    collection.cardMatrix[1][5] = cardA
    collection.cardMatrix[0][1] = cardB
    collection.cardMatrix[1][1] = cardB
    collection.cardMatrix[2][0] = cardC
    collection.cardMatrix[2][1] = cardC

    collection.collection_cards = [cardA, cardB, cardC]
    collection.confirmEdit = jest.fn()

    props = {
      collection,
      canEditCollection: true,
      gridW: 200,
      gridH: 200,
      gutter: 10,
      sortBy: 'order',
      selectedArea: { minX: null, minY: null, maxX: null, maxY: null },
      minX: null,
      loadCollectionCards: jest.fn(() => Promise.resolve()),
      updateCollection: jest.fn(),
      cardProperties: [],
      blankContentToolState: {},
      apiStore: fakeApiStore(),
      uiStore: fakeUiStore,
      routingStore: {
        routeTo: jest.fn(),
        push: jest.fn(),
      },
    }
    rerender = () => {
      props.collection.API_batchUpdateCardsWithUndo.mockClear()
      wrapper = shallow(<FoamcoreGrid.wrappedComponent {...props} />)
      component = wrapper.instance()
    }
    rerender()
    cards = props.collection.collection_cards
    component.gridRef = { scrollLeft: 0, scrollTop: 0 }
  })

  it('renders MovableGridCards', () => {
    expect(wrapper.find('MovableGridCard').length).toEqual(3)
  })

  describe('findOverlap', () => {
    it('finds filledSpot (or not) where a card is trying to be dragged', () => {
      // similar to calculateFilledSpots, but given a card (needs width and height >= 1)
      let fakeCard = { row: 1, col: 5, width: 1, height: 1 }
      let overlap = component.findOverlap(fakeCard)
      expect(overlap.card).toEqual(cardA)
      // 2x2 should stick out and overlap cardA
      fakeCard = { row: 0, col: 4, width: 2, height: 2 }
      overlap = component.findOverlap(fakeCard)
      expect(overlap.card).toEqual(cardA)
    })
  })

  describe('onDragStart', () => {
    describe('when dragging multiple cards', () => {
      let fakeDragMap, cardId, card
      beforeEach(() => {
        card = cards[0]
        cardId = card.id
        fakeDragMap = [
          {
            card,
            col: 0,
            row: 0,
          },
          {
            card: cards[1],
            col: 1,
            row: 0,
          },
        ]
        component.draggingMap = []
        component.originalCard = jest.fn().mockReturnValue(card)
        component.determineDragMap = jest.fn().mockReturnValue(fakeDragMap)
        component.onDragStart(cardId)
      })

      it('should determine the drag map with card id', () => {
        expect(component.originalCard).toHaveBeenCalledWith(cardId)
        expect(component.determineDragMap).toHaveBeenCalledWith(cardId)
      })

      it('should set the dragging map if dragging multiple cards', () => {
        expect(component.draggingMap).toEqual(fakeDragMap)
      })
    })
  })

  describe('resetCardPositions', () => {
    beforeEach(() => {
      component.moveCards = jest.fn().mockReturnValue()
      component.resizeCard = jest.fn().mockReturnValue()
      component.dragging = true
      props.uiStore.multiMoveCardIds = [cards[0].id]
    })

    it('should stop all dragging', () => {
      component.resetCardPositions()
      expect(component.dragGridSpot.size).toEqual(0)
      expect(component.dragging).toEqual(false)
      expect(props.uiStore.setMovingCards).toHaveBeenCalledWith([])
    })
  })

  describe('coordinatesForPosition', () => {
    it('should calculate the appropriate coordinates', () => {
      const { gridW, gutter } = v.defaultGridSettings
      const zoom = component.relativeZoomLevel
      let x = (gridW + gutter) / zoom
      const y = 0
      let width = 1
      expect(component.coordinatesForPosition({ x, y, width })).toEqual({
        col: 1,
        outsideDraggableArea: false,
        row: 0,
      })
      x = (16 * (gridW + gutter)) / zoom
      width = 4
      expect(component.coordinatesForPosition({ x, y, width })).toEqual({
        // this will get bumped back to 12 (where it fits)
        col: 12,
        outsideDraggableArea: true,
        row: 0,
      })
    })
  })

  describe('onDragOrResizeStop', () => {
    let cardId

    beforeEach(() => {
      cardId = cards[0].id
      component.moveCards = jest.fn().mockReturnValue()
      component.resizeCard = jest.fn().mockReturnValue()
      component.dragging = true
      props.uiStore.multiMoveCardIds = [cards[0].id]
      props.uiStore.selectedCardIds = [cards[0].id]
    })

    describe('when moving', () => {
      beforeEach(() => {
        component.onDragOrResizeStop(cardId, 'move')
      })

      it('calls moveCards', () => {
        expect(component.moveCards).toHaveBeenCalledWith(cards[0])
      })
    })

    describe('when resizing', () => {
      beforeEach(() => {
        component.resizing = true
        component.onDragOrResizeStop(cardId, 'resize')
      })

      it('calls resizeCard', () => {
        expect(component.resizeCard).toHaveBeenCalledWith(cards[0])
      })
    })
  })

  describe('onResize', () => {
    let cardId

    beforeEach(() => {
      component.resizing = false
      cardId = cards[0].id
      component.throttledSetResizeSpot = jest.fn().mockReturnValue()
      component.onResize(cardId, { width: 2, height: 1 })
    })

    it('should set resizing to true', () => {
      expect(component.resizing).toBe(true)
    })

    it('should call to set the resize spot with dimensions and position', () => {
      expect(component.throttledSetResizeSpot).toHaveBeenCalledWith({
        row: cards[0].row,
        col: cards[0].col,
        width: 2,
        height: 1,
      })
    })
  })

  describe('resizeCard', () => {
    beforeEach(() => {
      component.placeholderSpot = { width: 2, height: 2 }
    })

    it('calls collection.API_batchUpdateCardsWithUndo', () => {
      component.resizeCard(cards[0])
      expect(
        props.collection.API_batchUpdateCardsWithUndo
      ).toHaveBeenCalledWith({
        updates: [
          {
            card: cards[0],
            width: 2,
            height: 2,
          },
        ],
        undoMessage: 'Card resize undone',
        onConfirm: expect.any(Function),
      })
    })

    it('calls resetCardPositions', () => {
      component.resetCardPositions = jest.fn()
      component.resizeCard(cards[0])
      expect(component.resetCardPositions).toHaveBeenCalled()
    })
  })

  describe('moveCards', () => {
    const card = { id: '1', row: 0, col: 0, record: { internalType: 'items' } }
    const card2 = { ...card, id: '2' }
    describe('when moving a single card', () => {
      beforeEach(() => {
        component.dragGridSpot.set('6,7', { col: 6, row: 7, card })
        component.moveCards(card)
      })

      it('calls collection.API_batchUpdateCardsWithUndo', () => {
        expect(
          props.collection.API_batchUpdateCardsWithUndo
        ).toHaveBeenCalledWith({
          updates: [
            {
              card,
              col: 6,
              row: 7,
            },
          ],
          undoMessage: 'Card move undone',
          onConfirm: expect.any(Function),
          onCancel: expect.any(Function),
        })
      })
    })

    describe('when moving multiple cards', () => {
      beforeEach(() => {
        props.uiStore.multiMoveCardIds = [card.id, card2.id]
        rerender()
        component.dragGridSpot.set('6,7', { col: 6, row: 7, card })
        component.dragGridSpot.set('8,9', { col: 8, row: 9, card: card2 })
        component.draggingMap = [
          { card, row: 6, col: 7 },
          { card: card2, row: 8, col: 9 },
        ]
        component.moveCards(cards[0])
      })

      it('calls collection.API_batchUpdateCardsWithUndo', () => {
        expect(
          props.collection.API_batchUpdateCardsWithUndo
        ).toHaveBeenCalledWith({
          updates: [
            {
              card,
              col: 6,
              row: 7,
            },
            {
              card: card2,
              col: 8,
              row: 9,
            },
          ],
          undoMessage: 'Card move undone',
          onConfirm: expect.any(Function),
          onCancel: expect.any(Function),
        })
      })
    })

    describe('when dragging from MDL', () => {
      beforeEach(() => {
        CardMoveService.moveCards = jest.fn()
        props.uiStore.draggingFromMDL = true
        component.movingCards = [card, card2]
        component.dragGridSpot.set('6,7', { col: 6, row: 7, card })
        component.moveCards(card)
      })

      it('calls CardMoveService with the drag spot row/col', () => {
        expect(CardMoveService.moveCards).toHaveBeenCalledWith(
          {
            col: 6,
            row: 7,
          },
          { collection_card_ids: [] },
          // should pass in card as the "topLeftCard"
          card
        )
      })
    })

    describe('when the MoveSnackbar is open', () => {
      beforeEach(() => {
        props.uiStore.movingCardIds = [card.id, card2.id]
        props.apiStore = fakeApiStore({ findResult: card })
        rerender()
      })

      afterEach(() => {
        props.uiStore.movingCardIds = []
        props.apiStore.find.mockClear()
        CollectionCard.mockClear()
      })

      it('renders the MDL placeholder', () => {
        // 3 collection_cards + 1 MDL
        expect(wrapper.find('MovableGridCard').length).toEqual(4)
        // it creates a new CollectionCard
        expect(CollectionCard).toHaveBeenCalledTimes(1)
        const placeholderId = `${card.id}-mdlPlaceholder`
        expect(props.apiStore.updateModelId).toHaveBeenCalledWith(
          expect.any(Object),
          placeholderId
        )
        expect(wrapper.find('MovableGridCard').get(0).props.card.id).toEqual(
          placeholderId
        )
      })
    })
  })

  describe('calcEdgeCol/Row', () => {
    beforeEach(() => {
      cardA = createCard({ row: 1, col: 1 })
      cardB = createCard({ row: 1, col: 8 })
      cardC = createCard({ row: 1, col: 9 })
      const { collection } = props
      collection.cardMatrix = [[], [], [], []]
      collection.cardMatrix[1][1] = cardA
      collection.cardMatrix[1][8] = cardB
      collection.cardMatrix[1][9] = cardC

      props.collection.collection_cards = [cardA, cardB, cardC]
    })

    afterEach(() => {
      props.collection.cardMatrix = [[], [], [], []]
    })

    describe('with a card that has no cards around it', () => {
      it('should set the max column as the max card width', () => {
        const edgeCol = component.calcEdgeCol(cardA, cardA.id)
        expect(edgeCol).toEqual(4)
      })

      it('sets the max row as the max card height', () => {
        const edgeRow = component.calcEdgeRow(cardA, cardA.id)
        expect(edgeRow).toEqual(2)
      })
    })

    describe('with a card that has horizontal constraints, 2 spaces apart', () => {
      beforeEach(() => {
        cardB.row = 1
        cardB.col = 3
        const { collection } = props
        collection.cardMatrix[1][1] = cardA
        collection.cardMatrix[1][3] = cardB
        collection.cardMatrix[1][9] = cardC
      })

      it('has a maxResizeCol of 2', () => {
        const edgeCol = component.calcEdgeCol(cardA, cardA.id)
        expect(edgeCol).toEqual(2)
      })
    })

    describe('with a card that has vertical constraints, 1 space apart', () => {
      beforeEach(() => {
        cardA.row = 2
        cardB.row = 3
        cardB.col = 1
        const { collection } = props
        collection.cardMatrix[2][1] = cardA
        collection.cardMatrix[3][1] = cardB
        collection.cardMatrix[1][9] = cardC
      })

      it('has maxResizeRow of 1', () => {
        const edgeRow = component.calcEdgeRow(cardA, cardA.id)
        expect(edgeRow).toEqual(1)
      })
    })
  })

  describe('loadAfterScroll', () => {
    beforeEach(() => {
      const { collection, uiStore } = props
      collection.loadedRows = 9
      collection.loadedCols = 9
      component.loadCards = jest.fn()
      // zoomed out one level (this affects visibleRows)
      uiStore.zoomLevel = 2
    })

    describe('scrolling in loaded bounds', () => {
      beforeEach(() => {
        component.visibleCols = { min: 0, max: 4, num: 5 }
        component.visibleRows = { min: 1, max: 4, num: 4 }
      })

      it('does not call loadCards if all in view', () => {
        component.loadAfterScroll()
        expect(component.loadCards).not.toHaveBeenCalled()
      })
    })

    describe('scrolling out of bounds vertically', () => {
      it('calls loadMoreRows', () => {
        component.computeVisibleRows()
        component.loadAfterScroll()
        const minRow = props.collection.loadedRows + 1
        const expectedRows = {
          // ceil needed because visibleRows.num may be fractional
          rows: [minRow, Math.ceil(minRow + component.visibleRows.num + 3)],
        }
        expect(props.loadCollectionCards).toHaveBeenCalledWith(expectedRows)
      })
    })
  })

  describe('updateSelectedArea', () => {
    beforeEach(() => {
      cardA = createCard({ row: 0, col: 1 })
      cardB = createCard({ row: 1, col: 2 })
      cardC = createCard({ row: 3, col: 3 })
      props.collection.collection_cards = [cardA, cardB, cardC]
    })

    describe('selected area not matching any cards', () => {
      beforeEach(() => {
        props.collection.cardIdsWithinRectangle = jest.fn().mockReturnValue([])
        rerender()
        props.uiStore.selectedCardIds = []
        props.selectedArea = { minX: 500, minY: 10, maxX: 550, maxY: 20 }
        // It would be nice if we could use the real Collection class
        // instead of having to mock the return value:
        component.componentDidUpdate(props)
      })

      it('does not set uiStore.selectedCardIds', () => {
        expect(props.uiStore.selectedCardIds).toEqual([])
      })
    })

    describe('selected area matching two cards', () => {
      beforeEach(() => {
        props.selectedArea = { minX: 40, minY: 150, maxX: 550, maxY: 450 }
        // It would be nice if we could use the real Collection class
        // instead of having to mock the return value:
        props.collection.cardIdsWithinRectangle = jest
          .fn()
          .mockReturnValue([cardA.id, cardB.id])
        rerender()
        component.componentDidUpdate(props)
      })

      it('sets uiStore.selectedCardIds', () => {
        expect(props.uiStore.reselectCardIds).toHaveBeenCalledWith([
          cardA.id,
          cardB.id,
        ])
      })
    })
  })

  describe('maxCols', () => {
    describe('on touchDevice', () => {
      beforeEach(() => {
        props.collection.num_columns = 16
        props.uiStore.isTouchDevice = true
        props.uiStore.isMobile = true
        rerender()
      })
      it('should return the minimum of num_columns and 8', () => {
        expect(component.maxCols).toEqual(8)
      })
    })
    describe('on desktop', () => {
      beforeEach(() => {
        props.collection.num_columns = 4
        props.uiStore.isTouchDevice = false
        props.uiStore.isMobile = false
        rerender()
      })
      it('should return the minimum of num_columns and 16', () => {
        expect(component.maxCols).toEqual(4)
      })
    })
  })

  describe('relativeZoomLevel', () => {
    beforeEach(() => {
      props.collection.num_columns = 16
      props.collection.maxZoom = 3
      props.uiStore.zoomLevel = 3
      rerender()
    })
    describe('when zoomed all the way out', () => {
      it('should return the ratio that shows the entire grid by default', () => {
        const maxGridWidth = 5472
        expect(component.maxGridWidth({ zoomLevel: 3 })).toEqual(maxGridWidth)
        expect(component.relativeZoomLevel).toEqual(
          maxGridWidth / jestInnerWidth
        )
      })
    })
    describe('when zoomed in', () => {
      it('should return the zoomLevel', () => {
        props.uiStore.zoomLevel = 2
        expect(component.relativeZoomLevel).toEqual(2)
      })
    })
  })

  describe('handleZoomIn/Out', () => {
    it('should call the respective uiStore function', () => {
      const { uiStore } = props
      component.handleZoomIn()
      expect(uiStore.zoomIn).toHaveBeenCalled()
      component.handleZoomOut()
      expect(uiStore.zoomOut).toHaveBeenCalled()
    })
  })

  describe('showZoomControls', () => {
    beforeEach(() => {
      props.collection.num_columns = 4
      props.collection.isFourWideBoard = true
      rerender()
    })
    it('should return true if the innerWidth < maxGridWidth at zoomLevel 1', () => {
      const maxGridWidth = 1384
      expect(component.maxGridWidth({ zoomLevel: 1 })).toEqual(maxGridWidth)
      expect(maxGridWidth > jestInnerWidth).toEqual(true)
      // 1384 > 1024 so this will be true
      expect(component.showZoomControls).toEqual(true)
    })
  })

  describe('with different board sizes', () => {
    beforeEach(() => {
      props.collection.num_columns = 4
      props.collection.isFourWideBoard = true
      rerender()
    })
    it('should call uiStore.adjustZoomLevel to ensure zoom is correct', () => {
      expect(props.uiStore.adjustZoomLevel).toHaveBeenCalledWith({
        collection: props.collection,
      })
    })
  })
})
