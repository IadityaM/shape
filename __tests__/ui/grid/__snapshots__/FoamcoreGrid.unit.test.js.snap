// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FoamcoreGrid renders snapshot 1`] = `
<styled.div
  data-empty-space-click={true}
  height={NaN}
  innerRef={[Function]}
  onScroll={[Function]}
  width={1011.8095238095239}
>
  <FoamcoreZoomControls
    onZoomIn={[Function]}
    onZoomOut={[Function]}
  />
  <styled.div
    blocked={false}
    data-blank-type="unrendered"
    data-empty-space-click={true}
    draggedOn={true}
    h={250}
    height={250}
    key="blank-unrendered-1:5"
    onClick={[Function]}
    type="unrendered"
    w={312}
    width={312}
    x={316.1904761904762}
    xPos={316.1904761904762}
    y={51.42857142857143}
    yPos={51.42857142857143}
    zoomLevel={5.25}
  />
  <styled.div
    blocked={false}
    data-blank-type="unrendered"
    data-empty-space-click={true}
    draggedOn={true}
    h={520}
    height={520}
    key="blank-unrendered-0:1"
    onClick={[Function]}
    type="unrendered"
    w={644}
    width={644}
    x={63.23809523809524}
    xPos={63.23809523809524}
    y={0}
    yPos={0}
    zoomLevel={5.25}
  />
  <styled.div
    blocked={false}
    data-blank-type="unrendered"
    data-empty-space-click={true}
    draggedOn={true}
    h={250}
    height={250}
    key="blank-unrendered-2:0"
    onClick={[Function]}
    type="unrendered"
    w={644}
    width={644}
    x={0}
    xPos={0}
    y={102.85714285714286}
    yPos={102.85714285714286}
    zoomLevel={5.25}
  />
</styled.div>
`;
